File: android/app/src/main/kotlin/com/vyooo/insta360/Insta360FrameSink.kt
package com.vyooo.insta360

import android.util.Log
import com.arashivision.graphicpath.insmedia.common.MediaFrame
import com.vyooo.insta360.pipeline.DownscaleStage
import com.vyooo.insta360.pipeline.ForwardMaskStage
import com.vyooo.insta360.pipeline.FramePipeline
import com.vyooo.insta360.pipeline.HeuristicDecisionLayer
import com.vyooo.insta360.pipeline.MutableHints
import com.vyooo.insta360.pipeline.PanoramaDetectStage
import com.vyooo.insta360.pipeline.PipelineFrame
import com.vyooo.insta360.pipeline.TemporalDedupStage

/**
 * The single insertion point **and** processing hub for extracted Insta360 frames.
 *
 * One deterministic processing path:
 *
 * ```
 * MediaFrame (YUV420P) → YUV→RGB → FramePipeline[ Downscale → PanoramaDetect → ForwardMask →
 *                                                  TemporalDedup ] → processed RGBA
 *      ├─► onProcessedFrame  (host display — the Flutter texture; always)
 *      └─► onFrame           (encoder / transport — Agora; when streamingEnabled)
 * ```
 *
 * A stage may drop a frame (pipeline returns null) — nothing is forwarded for that frame, so the
 * display and the transmitted stream stay in lock-step. [Insta360GlRenderer] only uploads the
 * processed RGBA to the texture; the pipeline is the single source of truth for the optimisation.
 *
 * Frames arrive on the SDK extract thread; listeners marshal to the main thread as needed.
 */
object Insta360FrameSink {

    /** Processed RGBA for the host display. (rgba, w, h, ptsUs). The buffer is reused — read it now. */
    @Volatile var onProcessedFrame: ((ByteArray, Int, Int, Long) -> Unit)? = null

    /** Processed RGBA for the encoder/transport. Only when [streamingEnabled]; receives its own copy. */
    @Volatile var onFrame: ((ByteArray, Int, Int, Long) -> Unit)? = null

    /** (width, height, fps, totalCount) — emitted ~1×/sec. */
    @Volatile var onStats: ((Int, Int, Int, Long) -> Unit)? = null

    @Volatile var streamingEnabled: Boolean = false

    // ── The capture-side optimisation pipeline (single source of truth) ──────────
    private val forwardMask = ForwardMaskStage()
    private val temporalDedup = TemporalDedupStage()

    /** AI-fed decision hints; deterministic (all null) until the decision layer writes them. */
    val hints = MutableHints()

    /** M3 bounded heuristic decision layer — writes decision signals into [hints] (metadata only). */
    val decisionLayer = HeuristicDecisionLayer(hints)

    val pipeline = FramePipeline(
        listOf(DownscaleStage(), PanoramaDetectStage(), forwardMask, temporalDedup),
        hints,
    )

    /** Live masked/unmasked toggle (forward-only suppression on/off). */
    fun setMaskEnabled(enabled: Boolean) {
        forwardMask.enabled = enabled
    }

    /** Live temporal-reduction toggle (1-in-N + motion gating on/off) — for A/B / KPI capture. */
    fun setTemporalEnabled(enabled: Boolean) {
        temporalDedup.enabled = enabled
    }

    /** Live AI decision-layer toggle (off = deterministic fall-open) — for A/B / KPI capture. */
    fun setAiEnabled(enabled: Boolean) {
        decisionLayer.enabled = enabled
    }

    private var count: Long = 0
    private var windowStartNs: Long = 0
    private var framesThisWindow: Int = 0
    private var lastFps: Int = 0
    private var basePtsNs: Long = 0
    private var lastTransmitNs: Long = 0
    private var lastTemporalLogNs: Long = 0
    private var lastLogSeen: Long = 0
    private var lastLogKept: Long = 0
    private var lastLogSpans: Long = 0

    private const val TAG = "Insta360FrameSink"

    // Cap the transmit copy rate. Each transmitted frame allocates a fresh ~7 MB copy (+ another
    // ~7 MB in the platform-channel codec). The effective transmit rate is bottlenecked by the
    // native→Dart platform channel (~11 fps for these 7 MB frames), well below this cap, so 24 fps
    // is just a safety ceiling — raising it does not raise fps (that needs a native push path).
    private const val TRANSMIT_MIN_INTERVAL_NS = 1_000_000_000L / 24

    @Synchronized
    fun reset() {
        count = 0; windowStartNs = 0; framesThisWindow = 0; lastFps = 0; basePtsNs = 0
        lastTransmitNs = 0
        lastTemporalLogNs = 0; lastLogSeen = 0; lastLogKept = 0; lastLogSpans = 0
        pipeline.metrics.reset()
        temporalDedup.reset()
        decisionLayer.reset()
    }

    /** Live metrics: output fps + the pipeline snapshot (per-stage latency, spatial + temporal). */
    fun metrics(): Map<String, Any> {
        val m = HashMap<String, Any>()
        m["fps"] = lastFps
        m["framesOut"] = count
        m.putAll(pipeline.metrics.snapshot())
        m.putAll(temporalDedup.stats())
        m.putAll(decisionLayer.stats())
        return m
    }

    /**
     * Throttled (~1/s) logcat monitor of the temporal stage + the M3 decision layer — the on-device
     * stability readout. `capFps` (frames arriving from the camera) staying steady = real-time
     * sustained / no backlog; `effFps` (frames kept) is the temporally-reduced output rate — in a
     * static scene it should settle at `staticFps`, in motion it tracks `capFps`.
     *
     * AI side: `aiMoving` is the decision handed to the pacer (so `lastMotion` reads 1.000/0.000 when
     * the layer is on — that is the decision, not a score); `aiActivity` is the raw score behind it
     * (use it to tune `motionEnter`/`motionExit`); `aiSpans/s` is the **gate-stability** KPI — it must
     * stay near 0 in a steady scene, since a flapping gate gives the encoder irregular frame spacing.
     */
    private fun maybeLogTemporal(nowNs: Long) {
        val dt = nowNs - lastTemporalLogNs
        if (dt < 1_000_000_000L) return
        val s = temporalDedup.stats()
        val seen = s["framesSeen"] as Long
        val kept = s["framesKept"] as Long
        val secs = dt / 1_000_000_000.0
        val capFps = (seen - lastLogSeen) / secs
        val effFps = (kept - lastLogKept) / secs
        lastTemporalLogNs = nowNs
        lastLogSeen = seen
        lastLogKept = kept
        val ai = decisionLayer.stats()
        // Gate-stability KPI: static→moving transitions per second. A steady scene should sit near 0;
        // a climbing rate means the motion gate is flapping → irregular frame spacing into the encoder.
        val spans = ai["aiMotionSpans"] as Long
        val spansPerSec = (spans - lastLogSpans) / secs
        lastLogSpans = spans
        Log.i(
            TAG,
            "temporal enabled=${s["temporalEnabled"]} " +
                "capFps=${"%.1f".format(capFps)} effFps=${"%.1f".format(effFps)} " +
                "keepRatio=${"%.2f".format(s["keepRatio"] as Double)} " +
                "staticFps=${s["staticFps"]} " +
                "kept=$kept/$seen motionKeep=${s["motionKeeps"]} staticKeep=${s["staticKeeps"]} " +
                "staticDrop=${s["staticDrops"]} " +
                "lastMotion=${"%.3f".format(s["lastMotion"] as Float)} | " +
                "ai=${ai["aiEnabled"]} aiMoving=${ai["aiMoving"]} " +
                "aiActivity=${"%.3f".format(ai["aiActivity"] as Float)} " +
                "aiDetail=${"%.3f".format(ai["aiSpatialDetail"] as Float)} " +
                "aiRecScale=${"%.2f".format(ai["aiRecommendedScale"] as Float)} " +
                "aiTheta=${"%.1f".format(ai["aiThetaDeg"] as Float)} " +
                "aiSpans=$spans (${"%.1f".format(spansPerSec)}/s) " +
                "aiDecisions=${ai["aiDecisions"]} " +
                "aiMs=${"%.2f".format(ai["aiOverheadMs"] as Double)}",
        )
    }

    /**
     * Submit one extracted I420 [frame]: convert → run the pipeline → fan out to display + encoder.
     * Returns early (nothing forwarded) if conversion fails or a stage drops the frame.
     */
    @Synchronized
    fun submit(frame: MediaFrame) {
        val w = frame.width
        val h = frame.height
        if (w <= 0 || h <= 0) return

        val rgba = Insta360YuvConverter.toRgba(frame) ?: return

        val now = System.nanoTime()
        if (basePtsNs == 0L) basePtsNs = now
        val ptsUs = (now - basePtsNs) / 1000

        // M3: the heuristic decision layer observes this frame and writes decision signals (metadata)
        // into `hints` BEFORE the pipeline runs, so the stages read fresh AI-assisted hints this frame.
        decisionLayer.observe(rgba, w, h)

        val result = pipeline.process(PipelineFrame(rgba, w, h, ptsUs))
        maybeLogTemporal(now) // ~1/s monitor of the temporal stage — logs even on dropped frames
        result ?: return // dropped by a stage (e.g. temporal dedup)
        val outW = result.width
        val outH = result.height

        count++
        // fps over a 1-second sliding window
        if (windowStartNs == 0L) windowStartNs = now
        framesThisWindow++
        if (now - windowStartNs >= 1_000_000_000L) {
            lastFps = framesThisWindow
            framesThisWindow = 0
            windowStartNs = now
            onStats?.invoke(outW, outH, lastFps, count)
        }

        // Host display — reads the (reused) buffer immediately.
        onProcessedFrame?.invoke(result.pixels, outW, outH, ptsUs)

        // Encoder / transport — own copy so the consumer may retain it. Rate-capped (see
        // TRANSMIT_MIN_INTERVAL_NS): the 60 fps pipeline feeds the display, but transmitting every
        // frame allocated a fresh ~7 MB copy 60×/s and OOM-crashed the app under load.
        if (streamingEnabled && now - lastTransmitNs >= TRANSMIT_MIN_INTERVAL_NS) {
            val cb = onFrame
            if (cb != null) {
                lastTransmitNs = now
                val outBytes = outW * outH * 4
                val copy = ByteArray(outBytes)
                System.arraycopy(result.pixels, 0, copy, 0, outBytes)
                cb.invoke(copy, outW, outH, ptsUs)
            }
        }
    }
}

File: ios/Runner/Insta360/Insta360FrameSink.swift
import Foundation
import CoreVideo

/// The single insertion point + processing hub for extracted Insta360 flat-pano (ERP) frames.
/// iOS mirror of the Android `Insta360FrameSink` object — now running the SAME capture-side
/// optimisation pipeline (Downscale → PanoramaDetect → ForwardMask → TemporalDedup, fed by the
/// heuristic AI decision layer).
///
/// One deterministic processing path:
/// ```
/// BGRA CVPixelBuffer → RGBA → HeuristicDecisionLayer.observe → FramePipeline
///      ├─► onProcessedFrame (host display — BGRA CVPixelBuffer for the Flutter texture)
///      └─► onFrame          (encoder / transport — RGBA; when streamingEnabled)
/// ```
/// A stage may drop a frame (pipeline returns nil) — nothing is forwarded for that frame, so display
/// and transmitted stream stay in lock-step (identical to Android).
final class Insta360FrameSink {

    static let shared = Insta360FrameSink()

    // ── Pipeline (single source of truth, mirrors Android Insta360FrameSink) ─────
    private let downscale = DownscaleStage()
    private let forwardMask = ForwardMaskStage()
    private let temporalDedup = TemporalDedupStage()
    /// AI-fed decision hints; deterministic (all nil) until the decision layer writes them.
    let hints = MutableHints()
    /// M3 bounded heuristic decision layer — writes decision signals into `hints` (metadata only).
    let decisionLayer: HeuristicDecisionLayer
    let pipeline: FramePipeline

    private let pixelFactory = PixelBufferFactory()

    private init() {
        decisionLayer = HeuristicDecisionLayer(hints: hints)
        pipeline = FramePipeline(
            stages: [downscale, PanoramaDetectStage(), forwardMask, temporalDedup],
            hints: hints
        )
    }

    private let lock = NSLock()

    // ── Fan-out sinks (set by the bridge / texture) ─────────────────────────────
    /// Processed BGRA pixel buffer for the host texture.
    var onProcessedFrame: ((CVPixelBuffer, Int64) -> Void)?
    /// Processed RGBA (w*h*4) for the encoder/transport. Only invoked while `streamingEnabled`.
    var onFrame: ((Data, Int, Int, Int64) -> Void)?
    /// (width, height, fps, totalCount) — emitted ~1×/sec.
    var onStats: ((Int, Int, Int, Int64) -> Void)?

    var streamingEnabled = false

    // ── Pipeline toggles (now wired to the real stages) ─────────────────────────
    /// Live masked/unmasked toggle (forward-only suppression on/off).
    func setMaskEnabled(_ v: Bool) { forwardMask.enabled = v }
    /// Live temporal-reduction toggle (motion-gated static pacing on/off) — A/B / KPI capture.
    func setTemporalEnabled(_ v: Bool) { temporalDedup.enabled = v }
    /// Live AI decision-layer toggle (off = deterministic fall-open) — A/B / KPI capture.
    func setAiEnabled(_ v: Bool) { decisionLayer.enabled = v }

    // ── Stats window (mirrors Android FrameSink; drives the frameStats event) ────
    private var count: Int64 = 0
    private var windowStartNs: UInt64 = 0
    private var framesThisWindow = 0
    private var lastFps = 0
    private var baseTimestamp: TimeInterval = 0
    private var lastTransmitNs: UInt64 = 0

    /// Cap the transmit copy rate (mirrors Android's 24 fps ceiling).
    private let transmitMinIntervalNs: UInt64 = 1_000_000_000 / 24

    func reset() {
        lock.lock()
        count = 0; windowStartNs = 0; framesThisWindow = 0; lastFps = 0
        baseTimestamp = 0; lastTransmitNs = 0
        lock.unlock()
        pipeline.metrics.reset()
        downscale.reset()
        temporalDedup.reset()
        decisionLayer.reset()
    }

    /// Live metrics: FrameSink fps/framesOut, then the real pipeline snapshot + per-stage stats.
    /// Emits the identical keys Android does — `fps, framesIn, framesOut, framesDropped, totalMs,
    /// spatialReduction, stagesMs` (from the pipeline), plus the Downscale/Temporal/AI KPI fields.
    func metrics() -> [String: Any] {
        lock.lock()
        let fps = lastFps
        let framesOut = count
        lock.unlock()
        var m: [String: Any] = [:]
        m["fps"] = fps
        m["framesOut"] = framesOut
        m.merge(pipeline.metrics.snapshot()) { _, new in new }
        m.merge(downscale.stats()) { _, new in new }
        m.merge(temporalDedup.stats()) { _, new in new }
        m.merge(decisionLayer.stats()) { _, new in new }
        return m
    }

    /// Submit one stitched ERP frame. `pixelBuffer` is BGRA (`kCVPixelFormatType_32BGRA`).
    /// `timestamp` is the SDK video-frame timestamp (seconds). Called serially on the flat-pano queue.
    func submit(pixelBuffer: CVPixelBuffer, timestamp: TimeInterval) {
        let w = CVPixelBufferGetWidth(pixelBuffer)
        let h = CVPixelBufferGetHeight(pixelBuffer)
        guard w > 0, h > 0 else { return }
        guard var rgba = Self.bgraToRgba(pixelBuffer) else { return }

        let now = DispatchTime.now().uptimeNanoseconds
        lock.lock()
        if baseTimestamp == 0 { baseTimestamp = timestamp }
        let ptsUs = Int64((timestamp - baseTimestamp) * 1_000_000)
        lock.unlock()

        // M3: the decision layer observes this frame and writes decision signals (metadata) into
        // `hints` BEFORE the pipeline runs, so the stages read fresh AI-assisted hints this frame.
        decisionLayer.observe(rgba, width: w, height: h)

        let frame = PipelineFrame(pixels: rgba, width: w, height: h, ptsUs: ptsUs)
        // Drop this local's reference to the pixel storage so `frame.pixels` is uniquely owned — the
        // in-place ForwardMask mutation then avoids a copy-on-write copy at the full tier.
        rgba = []
        guard let result = pipeline.process(frame) else {
            return // dropped by a stage (e.g. temporal dedup) — metrics already recorded the drop
        }
        let outW = result.width
        let outH = result.height

        // fps over a 1-second sliding window (post-pipeline output rate).
        lock.lock()
        count += 1
        if windowStartNs == 0 { windowStartNs = now }
        framesThisWindow += 1
        var emitStats: (Int, Int, Int, Int64)?
        if now - windowStartNs >= 1_000_000_000 {
            lastFps = framesThisWindow
            framesThisWindow = 0
            windowStartNs = now
            emitStats = (outW, outH, lastFps, count)
        }
        let streaming = streamingEnabled
        let shouldTransmit = streaming && (now - lastTransmitNs >= transmitMinIntervalNs)
        if shouldTransmit { lastTransmitNs = now }
        let statsCb = onStats
        let processedCb = onProcessedFrame
        let frameCb = onFrame
        lock.unlock()

        // Host display — processed RGBA → BGRA pixel buffer for the Flutter texture.
        if let processedCb = processedCb,
           let pb = pixelFactory.makeBGRA(fromRGBA: result.pixels, width: outW, height: outH) {
            processedCb(pb, ptsUs)
        }

        if let s = emitStats { statsCb?(s.0, s.1, s.2, s.3) }

        // Encoder/transport — its own tightly-packed RGBA copy (only the first outW*outH*4 bytes;
        // Downscale may hand back an oversized backing array). Rate-capped.
        if shouldTransmit, let cb = frameCb {
            let need = outW * outH * 4
            let data = result.pixels.withUnsafeBufferPointer { buf -> Data in
                Data(bytes: buf.baseAddress!, count: need)
            }
            cb(data, outW, outH, ptsUs)
        }
    }

    /// Convert a BGRA `CVPixelBuffer` to a tightly-packed RGBA `[UInt8]` (the pipeline works in RGBA).
    /// Handles row padding (`bytesPerRow` may exceed `width*4`).
    private static func bgraToRgba(_ pixelBuffer: CVPixelBuffer) -> [UInt8]? {
        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }
        guard let base = CVPixelBufferGetBaseAddress(pixelBuffer) else { return nil }
        let w = CVPixelBufferGetWidth(pixelBuffer)
        let h = CVPixelBufferGetHeight(pixelBuffer)
        let srcStride = CVPixelBufferGetBytesPerRow(pixelBuffer)
        let dstStride = w * 4
        var out = [UInt8](repeating: 0, count: dstStride * h)
        let src = base.assumingMemoryBound(to: UInt8.self)
        out.withUnsafeMutableBufferPointer { dbuf in
            guard let dst = dbuf.baseAddress else { return }
            for y in 0..<h {
                let srcRow = src + y * srcStride
                let dstRow = dst + y * dstStride
                var x = 0
                while x < w {
                    let i = x * 4
                    // BGRA -> RGBA: swap B and R, keep G and A.
                    dstRow[i + 0] = srcRow[i + 2]
                    dstRow[i + 1] = srcRow[i + 1]
                    dstRow[i + 2] = srcRow[i + 0]
                    dstRow[i + 3] = srcRow[i + 3]
                    x += 1
                }
            }
        }
        return out
    }
}
